packages/adblocker/src/engine/engine.ts

```diff
@@ -908,31 +1023,49 @@ export default class FilterEngine extends EventEmitter<
     }
 
     // Collect all CSP directives and keep track of exceptions
-    const disabledCsp = new Set();
-    const enabledCsp = new Set();
+    const exceptions: Map<string | undefined, NetworkFilter> = new Map();
+    const filters: NetworkFilter[] = [];
+
     for (const filter of matches) {
       if (filter.isException()) {
         if (filter.csp === undefined) {
           // All CSP directives are disabled for this site
+          this.emit(
+            'filter-matched',
+            { exception: filter },
+            { request, filterType: FilterType.NETWORK },
+          );
           return undefined;
         }
-        disabledCsp.add(filter.csp);
+        exceptions.set(filter.csp, filter);
       } else {
-        enabledCsp.add(filter.csp);
+        filters.push(filter);
       }
     }
 
+    if (filters.length === 0) {
+      return undefined;
+    }
+
+    const enabledCsp = new Set();
+
     // Combine all CSPs (except the black-listed ones)
-    const csps: string | undefined =
-      Array.from(enabledCsp)
-        .filter((csp) => !disabledCsp.has(csp))
-        .join('; ') || undefined;
-
-    // Emit event
-    if (csps !== undefined) {
-      this.emit('csp-injected', csps, request);
+    for (const filter of filters.values()) {
+      const exception = exceptions.get(filter.csp);
+      if (exception === undefined) {
+        enabledCsp.add(filter.csp);
+      }
+      this.emit(
+        'filter-matched',
+        { filter, exception },
+        { request, filterType: FilterType.NETWORK },
+      );
     }
 
+    const csps = Array.from(enabledCsp).join('; ');
+
+    this.emit('csp-injected', request, csps);
```

Comment:
Should this be omitted if `csps` is empty?

Reply:
good catch!